### PR TITLE
Graph[]

### PR DESCRIPTION
**Pull request by poke1024**
Still a work in progress, but already usable.

### CODE REVIEW

File: mathics/builtin/graphs.py [hunk 1]
```diff
@@ -0,0 +1,1301 @@
...
+                return Expression('List', *nx.minimum_node_cut(graph.G))
+
+    def apply_st(self, graph, s, t, evaluation, options):
+        'FindVertexCut[graph_, s_, t_, OptionsPattern[%(name)s]]'
+        graph = self._build_graph(graph, evaluation, options, lambda: Expression('FindVertexCut', graph, s, t))
+        if graph:
+            if not _is_connected(graph.G):
+                return Expression('List')
+            else:
+                return Expression('List', *nx.minimum_node_cut(graph.G, s, t))
```
Comment (sn6uv) [inline mathics/builtin/graphs.py]: Should `s` and `t` be converted to python?

File: mathics/builtin/graphs.py [hunk 2]
```diff
@@ -0,0 +1,1301 @@
...
+     = 1
+
+    >> EdgeConnectivity[{1 -> 2, 2 -> 3}]
+     = 0
+
+    >> EdgeConnectivity[{1 -> 2, 2 -> 3, 3 -> 1}]
+     = 1
+
+    >> EdgeConnectivity[{1 <-> 2, 2 <-> 3, 1 <-> 3}]
+     = 2
```
Comment (sn6uv) [inline mathics/builtin/graphs.py]: empty graph

File: mathics/builtin/graphs.py [hunk 3]
```diff
@@ -0,0 +1,1301 @@
...
+     = 1
+
+    >> VertexConnectivity[{1 -> 2, 2 -> 3}]
+     = 0
+
+    >> VertexConnectivity[{1 -> 2, 2 -> 3, 3 -> 1}]
+     = 1
+
+    >> VertexConnectivity[{1 <-> 2, 2 <-> 3, 1 <-> 3}]
+     = 2
```
Comment (sn6uv) [inline mathics/builtin/graphs.py]: empty graph

File: mathics/builtin/graphs.py [hunk 4]
```diff
@@ -0,0 +1,1301 @@
...
+                return Integer(nx.node_connectivity(graph.G))
+
+    def apply_st(self, graph, s, t, evaluation, options):
+        '%(name)s[graph_, s_, t_, OptionsPattern[%(name)s]]'
+        graph = self._build_graph(graph, evaluation, options, lambda: Expression('VertexConnectivity', graph, s, t))
+        if graph:
+            if not _is_connected(graph.G):
+                return Integer(0)
+            else:
+                return Integer(nx.node_connectivity(graph.G, s, t))
```
Comment (sn6uv) [inline mathics/builtin/graphs.py]: `s` and `t` need conversion?

File: mathics/builtin/graphs.py [hunk 5]
```diff
@@ -0,0 +1,1301 @@
...
+     = {1, 2, 3, 4, 5}
+
+    >> FindShortestPath[{1 <-> 2, 2 <-> 3, 4 -> 3, 4 -> 2, 4 -> 5}, 1, 5]
+     = {}
+
+    >> g = Graph[{1 -> 2, 2 -> 3, 1 -> 3}, EdgeWeight -> {0.5, a, 3}];
+    >> a = 0.5; FindShortestPath[g, 1, 3]
+     = {1, 2, 3}
+    >> a = 10; FindShortestPath[g, 1, 3]
+     = {1, 3}
```
Comment (sn6uv) [inline mathics/builtin/graphs.py]:
```
FindShortestPath[{}, 1, 2}]
FindShortestPath[{1 -> 2}, 1, 3]
```

File: mathics/builtin/graphs.py [hunk 6]
```diff
@@ -0,0 +1,1301 @@
...
+     = 4
+
+    >> GraphDistance[{1 <-> 2, 2 <-> 3, 4 -> 3, 4 -> 2, 4 -> 5}, 1, 5]
+     = Infinity
+
+    >> GraphDistance[{1 <-> 2, 2 <-> 3, 3 <-> 4, 2 <-> 4, 4 -> 5}, 3]
+     = {2, 1, 0, 1, 2}
+
+    >> GraphDistance[{1 <-> 2, 3 <-> 4}, 3]
+     = {Infinity, Infinity, 0, 1}
```
Comment (sn6uv) [inline mathics/builtin/graphs.py]:
```
GraphDistance[{}, 1, 1]  (* returns 0 should be an error *)
GraphDistance[{1 -> 2}, 3, 4]
```

File: mathics/builtin/graphs.py [hunk 7]
```diff
@@ -0,0 +1,1301 @@
...
+            except nx.exception.NetworkXNoPath:
+                return Expression('DirectedInfinity', 1)
+
+
+class CompleteGraph(_NetworkXBuiltin):
+    '''
+    >> CompleteGraph[8]
+     = -Graph-
+    '''
+
```
Comment (sn6uv) [inline mathics/builtin/graphs.py]: requires numpy

File: mathics/builtin/graphs.py [hunk 8]
```diff
@@ -0,0 +1,1301 @@
...
+    def apply(self, n, evaluation, options):
+        '%(name)s[n_Integer, OptionsPattern[%(name)s]]'
+        py_n = n.get_int_value()
```
Comment (sn6uv) [inline mathics/builtin/graphs.py]: check `py_n >= 1`

**Comment by sn6uv** [comment]
I'd tentatively like to suggest including this in 1.0 but if it's not ready then that's okay too.

File: mathics/builtin/graphs.py [hunk 9]
```diff
@@ -0,0 +1,1427 @@
...
+     = False
+
+    #> PathGraphQ[Graph[{}]]
+     = False
+    #> PathGraphQ[Graph[{1 -> 2, 3 -> 4}]]
+     = False
+    #> PathGraphQ[Graph[{1 -> 2, 2 -> 1}]]
+     = True
+    #> PathGraphQ[Graph[{}]]
+     = False
```
Comment (sn6uv) [inline mathics/builtin/graphs.py]: this test is doubled up

**Comment by sn6uv** [comment]
Note: Now that #604 is merged the graph tests should now run on Travis.

**Comment by sn6uv** [comment]
Let's leave this for the next release.

**Comment by GarkGarcia** [comment]
@poke1024 Hey, would you be willing to rebase this branch so that we can _finally_ merge it?

File: mathics/core/parser/operators.py [hunk 10]
```diff
@@ -61,6 +61,7 @@
     'SuchThat': 180,
     'Rule': 120,
     'RuleDelayed': 120,
+    'UndirectedEdge': 120,
```
[the remark this answers is not in the document]
Comment (poke1024) [reply mathics/core/parser/operators.py]: This has now been addressed, `UndirectedEdge` is in `nonassoc_binary_ops` now.

File: mathics/builtin/graphs.py [hunk 11]
```diff
@@ -0,0 +1,1301 @@
...
+        return new_graph, 'WEIGHT'
+
+
+def _is_path(vertices, G):
+    return all(d <= 2 for d in G.degree(vertices).values())
+
+
+def _is_connected(G):
+    if isinstance(G, (nx.MultiDiGraph, nx.DiGraph)):
+        return len(list(nx.strongly_connected_components(G))) == 1
```
[the remark this answers is not in the document]
Comment (poke1024) [reply mathics/builtin/graphs.py]: not addressed yet, might be an optimization for networkx actually

File: mathics/builtin/graphs.py [hunk 12]
```diff
@@ -0,0 +1,1301 @@
...
+
+    >> g = Graph[{1 -> 2, 2 -> 3, 3 -> 1}]; ConnectedGraphQ[g]
+     = True
+
+    >> g = Graph[{1 <-> 2, 2 <-> 3}]; ConnectedGraphQ[g]
+     = True
+
+    >> g = Graph[{1 <-> 2, 2 <-> 3, 4 <-> 5}]; ConnectedGraphQ[g]
+     = False
+    '''
```
[the remark this answers is not in the document]
Comment (poke1024) [reply mathics/builtin/graphs.py:1376]: added

File: mathics/builtin/graphs.py [hunk 13]
```diff
@@ -0,0 +1,1301 @@
...
+        graph = self._build_graph(graph, evaluation, options, lambda: Expression('WeaklyConnectedComponents', graph))
+        if graph:
+            vertices_sorted = graph.vertices_sorted()
+            return Expression(
+                'List',
+                *[Expression('List', *vertices_sorted(c)) for c in nx.connected_components(graph.G.to_undirected())])
+
+
+class FindVertexCut(_NetworkXBuiltin):
+    '''
```
[the remark this answers is not in the document]
Comment (poke1024) [reply mathics/builtin/graphs.py:1501]: added now

File: mathics/builtin/graphs.py (hunk 7 above)
Comment (poke1024) [reply mathics/builtin/graphs.py:2170]: I believe this has been fixed

File: mathics/builtin/graphs.py (hunk 8 above)
Comment (poke1024) [reply mathics/builtin/graphs.py:2177]: check added

File: mathics/builtin/graphs.py [hunk 14]
```diff
@@ -0,0 +1,1301 @@
...
+
+    edge_weights = _edge_weights(options)
+
+    class ParseError(Exception):
+        pass
+
+    def parse_edge(r, attr_dict):
+        name = r.get_head_name()
+
+        if name == 'System`Property' and len(r.leaves) == 2:
```
[the remark this answers is not in the document]
Comment (poke1024) [reply mathics/builtin/graphs.py]: changed accordingly.

File: mathics/builtin/graphs.py (hunk 6 above)
Comment (poke1024) [reply mathics/builtin/graphs.py:2123]: added test cases accordingly

**Comment by GarkGarcia** [comment]
@poke1024 Thanks for the rebase! It looks like most failing tests are caused by `AttributeError`s, any idea why?

**Comment by poke1024** [comment]
@GarkGarcia I'm currently looking into this.

**Comment by GarkGarcia** [comment]
Great! Thanks for your contribution!

**Pull request merged**